src/server.c

```
@@ -182,13 +197,82 @@ static inline int makeSocket(unsigned int port)
     return sock;
 }
 
-static inline void handle(Server *server, int fd, fd_set *activeFDs, struct sockaddr_in *addr)
+static void setFdNonblocking(int fd)
+{
+    if (fcntl(fd, F_SETFL, fcntl(fd, F_GETFL) | O_NONBLOCK) < 0)
+	perror("fcntl");
+}
+
+static void serverAddFd(int epollfd, int fd, int in, int oneshot)
+{
+#if defined(__linux__)
+    struct epoll_event event;
+    event.data.fd = fd;
+    event.events = EPOLLET | EPOLLRDHUP;
+    event.events |= in ? EPOLLIN : EPOLLOUT;
+    if (oneshot)
+        event.events |= EPOLLONESHOT;
+    if (epoll_ctl(epollfd, EPOLL_CTL_ADD, fd, &event) < 0)
+	    perror("epoll_ctl");
+#elif defined(__APPLE__)
+    struct kevent event;
+    event.ident = fd;
+    event.flags = EV_ADD | EV_ENABLE | EV_CLEAR;
+    if (oneshot)
+        event.flags |= EV_ONESHOT;
+    EV_SET(&event, fd, EVFILT_READ, event.flags, 0, 0, NULL);
+    if(kevent(epollfd, &event, 1, NULL, 0, NULL) < 0) {
+        perror("kevent");
+    }
+#endif
+    setFdNonblocking(fd);
+}
+
+static void resetOneShot(int epollfd, int fd)
+{
+#if defined(__linux__)
+    struct epoll_event event;
+    event.data.fd = fd;
+    event.events = EPOLLIN | EPOLLET | EPOLLRDHUP | EPOLLONESHOT;
+    if (epoll_ctl(epollfd, EPOLL_CTL_MOD, fd, &event) < 0)
+	    perror("epoll_ctl");
+#elif defined(__APPLE__)
+    struct kevent event;
+    event.ident = fd;
+    event.flags = EV_ADD | EV_ENABLE | EV_CLEAR | EV_ONESHOT;
+    EV_SET(&event, fd, EVFILT_READ, event.flags, 0, 0, NULL);
+    if(kevent(epollfd, &event, 1, NULL, 0, NULL) < 0) {
+        perror("kevent");
+    }
```

Add `#else` condition. You can raise the warnings about unsupported platforms.